### PR TITLE
WIP add AVRs to ArchiveBlock proto

### PR DESCRIPTION
**Pull request by garbageslam**
This is an exploratory branch based on the idea that, we should
store the AVRs that support the block signatures in Amazon S3 storage,
in addition to storing them with the watchers.

This PR adds the verification report to the ArchiveBlock object.

That change requires moving the `BlockData` object out of `mc-transaction-core`
because it isn't acceptable for `mc-transaction-core` to depend on
`mc-attest-core`, that could make it very difficult to create hardware
wallets and such without cross-compiling `mc-attest-core` and all
its dependencies.

It also isn't good for `mc-attest-core` to depend on `mc-transaction-core`.

We decided that we should think of `BlockData` as a ledger type, since
it's really an S3 protobuf type corresponding to the `ArchiveBlock`
in `external.proto`. We moved it to `mc-ledger-types` crate, and renamed
it to `ArchiveBlock` to match the protobuf.

We then made
- `mc-ledger-db` have storage for verification reports, and de-duplicate
  these reports.
- `mc-ledger-sync` propagates any found verification reports to callers
  etc.
- More such plumbing like this.

It's not clear that we needed to extend `mc-ledger-db` in this way
to meet the goal, but it seems like it doesn't hurt and we also wanted
to see how much work this is.

---

Feedback on the direction in this PR is greatly appreciated. What
we learned from developing this will also be helpful for trying to
implement Guaranteed Timestamps.

(Guaranteed Timestamps means roughly, the Ledger also tracks the
block signatures not just of its own node, but also of peer nodes,
when blocks are obtained from the catchup process. Those peer
signatures are also added to `ArchiveBlock`. They are stored
separately from the node's own signature to avoid ambiguity. The
benefit of this is that there will not be a scenario going forward
that you get blocks with no signature and no timestamp if you download
from S3 -- currently fog has to rely on the watcher for timestamps,
which is more complicated. That is currently scheduled as a Q4 item IIRC)

### CODE REVIEW

File: api/proto/blockchain.proto [hunk 1]
```diff
@@ -74,6 +74,9 @@ message ArchiveBlockV1 {
 
     // Block signature, when available.
     BlockSignature signature = 3;
+
+    // Attestation Verification Report, when available
+    external.VerificationReport verification_report = 4;
```
Comment (garbageslam) [inline api/proto/blockchain.proto:79]: Adding this is the main point of the PR

File: consensus/service/src/tx_manager/mod.rs [hunk 2]
```diff
@@ -295,7 +296,10 @@ impl<E: ConsensusEnclave + Send, UI: UntrustedInterfaces + Send> TxManager
         // The enclave cannot provide a timestamp, so this happens in untrusted.
         signature.set_signed_at(chrono::Utc::now().timestamp() as u64);
 
-        Ok((block, block_contents, signature))
+        // Get the report which underlies the signature
+        let verification_report = self.enclave.get_ias_report()?;
```
Comment (garbageslam) [inline consensus/service/src/tx_manager/mod.rs:300]: this is where we get the `verification_report` to put in the `ArchiveBlock`

File: ledger/db/src/lib.rs [hunk 3]
```diff
@@ -235,7 +264,7 @@ impl Ledger for LedgerDB {
             Err(err) => Err(err),
         }?;
 
-        Ok(BlockData::new(block, contents, signature))
+        Ok(ArchiveBlock::new(block, contents, signature, None))
```
Comment (eranrund) [inline ledger/db/src/lib.rs:267]: This feels hacky to me. If there's a signature and we have an AVR for the signer shouldn't we return it as well?

File: ledger/db/src/lib.rs [hunk 4]
```diff
@@ -538,6 +574,21 @@ impl LedgerDB {
...
+        report: &VerificationReport,
+        db_transaction: &mut RwTransaction,
+    ) -> Result<(), Error> {
+        let signer_bytes = get_signer_bytes(report)?;
+        let avr_bytes = encode(report);
+        Ok(db_transaction.put(
+            self.verification_reports_by_signer,
+            &signer_bytes,
+            &avr_bytes,
+            WriteFlags::empty(), // Allow overwrite, overwrite is not expected
```
Comment (eranrund) [inline ledger/db/src/lib.rs:588]: Not expected but would indicate something very fishy is going on...

**Review by eranrund — COMMENTED** [review]
I think this looks good, thank you for building it!

Some things to consider:
* The database version should be updated (see https://github.com/mobilecoinfoundation/mobilecoin/blob/master/ledger/db/src/lib.rs#L69) (this will lead to https://github.com/mobilecoinfoundation/mobilecoin/blob/d7f655e660789a2bf7842d528d3dff953939691a/util/lmdb/src/metadata_store.rs#L65 returning `VersionIncompatible` unless you override it).
* Has this been tested on an existing ledger db? You might need to add migration code to create the new DB. See https://github.com/mobilecoinfoundation/mobilecoin/blob/d7f655e660789a2bf7842d528d3dff953939691a/ledger/migration/src/main.rs. Probably don't need to do it there since its just creating an empty db if it doesn't exist
* This might be worth breaking into two PRs: one that just deals with the renaming and relocation of `BlockData` -> `ArchiveBlock` and then one that adds the AVR. Should make reviewing easier

**Comment by garbageslam** [comment]
Thanks Eran, I will work on splitting this

**Comment by remoun** [comment]
This is mostly superseded by #1823 -- that feature branch still needs LedgerDB updates, but the addition of `SignedBlockMetadata` makes `BlockData` and `ArchiveBlock` include AVRs and SCP QuorumSets.

**Pull request closed without merging**